### PR TITLE
Flink 5990

### PR DESCRIPTION
**Pull request by sunjincheng121**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    }
+    lastTriggerTimestamp = timestamp
+  }
+
+  /**
+    * Aggregate calculations for the current element
+    *
+    * @param timestamp timestamp of current element
+    * @param input     the element to be calculated
+    * @param out       the collector of calculate the results
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: the collector of the results

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+      if (bufferedRowList.isEmpty) {
+        val bufferPriorityIt = bufferOperatorState.get().iterator()
+        while (bufferPriorityIt.hasNext) {
+          val ts = bufferPriorityIt.next()
+          bufferedRowList.add(ts)
+        }
+      }
+    }
+
+    if (null == lastExpiredRow) {
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: 这个是什么情况下发生的？

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+  private var accumulators: Row = _
+  private var sameTimeDataBuffer: JList[Row] = _
+
+  // save last expired data
+  private var lastExpiredRow: Row = _
+  // persistent last expired data for recovery
+  private var lastExpiredRowState: ListState[Row] = null
+
+  // for event-time, sorted the data by timestamp
+  private var bufferedRowPriorityQueue: PriorityQueue[(Long, Long)] = _
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: 这个bufferedRowPriorityQueue太重要了，一定要好好解释
你写下中文或者英文，我看一下

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+
+    if (null == lastExpiredRow) {
+      val lastValueOperatorIt = lastExpiredRowState.get().iterator()
+      if (lastValueOperatorIt.hasNext) {
+        lastExpiredRow = lastValueOperatorIt.next()
+      }
+    }
+
+    if (isRowTimeType) {
+      // save current timestamp to buffered row priority-queue which manage window data
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: keep the timestamp to the priority-queue
这个priority-queue在上面第一次定义的时候一定要解释清楚，上面解释清楚了，这里简单解释一下操作就好了

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+      bufferedRowPriorityQueue.enqueue((timestamp, Long.MaxValue - timestamp))
+      if (bufferedRowPriorityQueue.size > processingOffset) {
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: processingOffset 是什么？
这些解释的不是很清楚
@param processingOffset     the offset of processing
val processingOffset = overWindow.lowerBound.getOffset.asInstanceOf[RexInputRef].getIndex

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+  private var bufferedRowList: JList[Long] = _
+  // persistent buffered data for recovery
+  private var bufferOperatorState: ListState[Long] = null
+
+
+  // save incremental calculation accumulator
+  private var accumulatorState: ValueState[Row] = _
+
+  // save offset window data records
+  private var windowDataState: MapState[Long, JList[Row]] = _
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: 这个是什么意思？ 用一个state存下每个window的offset？
如果是，用windowOffsetMap

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+
+  // for event-time, sorted the data by timestamp
+  private var bufferedRowPriorityQueue: PriorityQueue[(Long, Long)] = _
+  // for proc-time
+  private var bufferedRowList: JList[Long] = _
+  // persistent buffered data for recovery
+  private var bufferOperatorState: ListState[Long] = null
+
+
+  // save incremental calculation accumulator
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: the state to keep accumulator for incremental calculation

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+  // save last expired data
+  private var lastExpiredRow: Row = _
+  // persistent last expired data for recovery
+  private var lastExpiredRowState: ListState[Row] = null
+
+  // for event-time, sorted the data by timestamp
+  private var bufferedRowPriorityQueue: PriorityQueue[(Long, Long)] = _
+  // for proc-time
+  private var bufferedRowList: JList[Long] = _
+  // persistent buffered data for recovery
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: the state to keep operator（这个operator是什么）for failover recovery

Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: 建议对所有的state，都解释一下参数的意义

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+  Preconditions.checkNotNull(aggFields)
+  Preconditions.checkArgument(aggregates.length == aggFields.length)
+
+  private var output: Row = _
+  private var accumulators: Row = _
+  private var sameTimeDataBuffer: JList[Row] = _
+
+  // save last expired data
+  private var lastExpiredRow: Row = _
+  // persistent last expired data for recovery
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: The state which keeps the last expired data for failover recovery

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+
+  Preconditions.checkNotNull(aggregates)
+  Preconditions.checkNotNull(aggFields)
+  Preconditions.checkArgument(aggregates.length == aggFields.length)
+
+  private var output: Row = _
+  private var accumulators: Row = _
+  private var sameTimeDataBuffer: JList[Row] = _
+
+  // save last expired data
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: the intermediate buffer which keeps the last expired row

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala (hunk 3 above)
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: 用于保存一个窗口内所有元素的时间戳，时间戳是tuple的第一个元素，tuple的第二个元素是用于priorityQueue的排序（Long.Max - ts), 这个pq的主要作用是retract过期数据。

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala (hunk 8 above)
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: bufferedRowList，the intermediate buffer which keeps all the rows in one window

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala (hunk 7 above)
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: bufferOperatorState -》 TimestampState 如果是event-time是bufferedRowPriorityQueue，如果是proc-time是bufferedRowList

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala (hunk 8 above)
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: bufferedRowPriorityQueue timestampPriorityQueue
bufferedRowList -》 timestampList

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala (hunk 6 above)
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: 存放一个窗口内容的所有元素，tuple的第一个位置是元素的timestamp，tuple的第二元素是相同timestamp内的所有数据。
windowDataState -> dataState

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala (hunk 3 above)
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: A priority queue which keeps the time stamps of the data that are not expired. The first element of the tuple is the origin time stamp (ts) of each data. The second element of the tuple is the derived time stamp (Long.Max - ts), as we want a min-priority queue based on ts. This timeStampPriorityQueue is used for retracting the expired  (out of window scope) data while the over window is sliding.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala (hunk 6 above)
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: A backend state which keeps all the data that are not expired. The first element (as the mapState key) of the tuple is the time stamp. Per each time stamp, the second element of tuple is a list that contains the entire data of all the rows belonging to this time stamp.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala (hunk 7 above)
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RowsClauseBoundedOverProcessFunction.scala]: A backend state which is used to materialize the time stamp. For event-time processing, it saves the bufferedRowPriorityQueue, while for proc-time, it saves bufferedRowList.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala [hunk 11]
```diff
@@ -293,6 +300,130 @@ class SqlITCase extends StreamingWithStateTestBase {
...
+    // TODO when we support multi "OVER" window when doing well
+    val config = new JHashMap[String,String]()
+    // need some advice on the definition of this constant
+    config.put(OVER_EVENT_TIME_ALLOWED_LATENESS, "2")
+
+    env.getConfig.setGlobalJobParameters(ParameterTool.fromMap(config))
+
+    // make sure that element (20L, 20, "Hello World") arrived at processing node
+    // followed by element (9L, 9, "Hello World").
+    env.setParallelism(1)
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala]: set the parallelism to 1 such that the test elements are arrived in order. For instance, element (20L, 20, "Hello World") arrives before element (9L, 9, "Hello World").

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala [hunk 12]
```diff
@@ -293,6 +300,130 @@ class SqlITCase extends StreamingWithStateTestBase {
...
+      (8L, 8, "Hello World"),
+      (20L, 20, "Hello World"))
+
+    val env = StreamExecutionEnvironment.getExecutionEnvironment
+    env.setStateBackend(getStateBackend)
+    val tEnv = TableEnvironment.getTableEnvironment(env)
+    StreamITCase.clear
+
+    // for sum aggregation ensure that every time the order of each element is consistent
+    env.setParallelism(1)
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala]: same as above

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala [hunk 13]
```diff
@@ -317,4 +448,24 @@ class SqlITCase extends StreamingWithStateTestBase {
     result.addSink(new StreamITCase.StringSink)
     env.execute()
   }
+
+}
+
+object SqlITCase {
+
+  class TimestampWithLatenessWatermark() extends AssignerWithPunctuatedWatermarks[(Long,
+    Int, String)] {
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala]: format

**Pull request closed without merging**